packages/native-stack/src/views/RetainContext.tsx

```diff
@@ -0,0 +1,138 @@
+import type {
+  ParamListBase,
+  Route,
+  StackNavigationState,
+} from '@react-navigation/native';
+import * as React from 'react';
+
+import type {
+  NativeStackDescriptorMap,
+  NativeStackNavigationHelpers,
+} from '../types';
+
+type RetainContextT = {
+  retain(): string;
+  release(key: string): boolean;
+  restore(key: string): boolean;
+  supported: boolean;
+};
+
+/**
+ * This context allows you to retain screens in memory, even after
+ * they have been popped off the stack. This is important for
+ * supporting native iOS "Picture in Picture" mode, as the
+ * originating UIViewController must be retained in memory for the
+ * PiP overlay to continue running.
+ * Retained screens can also be "restored" to the top of the stack,
+ * for example when the user taps the button in the PiP overlay
+ * to return to the app.
+ */
+export const RetainContext = React.createContext<RetainContextT>({
+  retain() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  release() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  restore() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  supported: false,
+});
+
+type RetainedScenes = {
+  [key: string]: {
+    route: Route<any>;
+    descriptor: NativeStackDescriptorMap['key'];
+  };
+};
+
+export function useRetainContext(
+  state: StackNavigationState<ParamListBase>,
+  navigation: NativeStackNavigationHelpers,
+  descriptors: NativeStackDescriptorMap
+) {
+  const [retainedScenes, setRetainedScenes] = React.useState<RetainedScenes>(
+    {}
+  );
+  const latestValues = React.useRef({ state, retainedScenes, descriptors });
+  React.useEffect(() => {
+    latestValues.current = { state, retainedScenes, descriptors };
+  });
+
+  const retainContext = React.useMemo<RetainContextT>(
+    () => ({
+      retain() {
+        const { state, descriptors } = latestValues.current;
+        const route = state.routes[state.routes.length - 1];
+        setRetainedScenes((screens) => ({
+          ...screens,
+          [route.key]: {
+            route,
+            descriptor: descriptors[route.key],
+          },
```

Comment:
This should only store the route key. Duplicating the `route` and `descriptor` is problematic as those could change after storing here making them out of sync. They should always be read from `state.routes` and `descriptors`.

Reply:
Hmm it can't be read from `state.routes` if the screen has been popped off the stack; that's why I'm storing it here. In the case where the route hasn't been popped, I use the one from `state.routes`.